### PR TITLE
NF core reassignment on termination

### PR DESCRIPTION
**Pull request by koolzz**
Adds functionality to onvm_mgr to reassign cores on NF shutdown.

## Summary:
Reallocates a NF to another core when another NF shuts down and a better core allocation is possible.
This is a part of multiple changes regarding the autoscaling and core managing improvments for ONVM, working with @ratnadeepb on these.

**Usage:**
Run onvm_mgr with 2 cores for NFs, f.e:
` ./go.sh 0,1,2 0 0x60`
Run basic monitor (or any other NF really)
`./go.sh 2`
Run basic monitor in shared more  (or any other NF really)
`./go.sh -l 0 -- -r 1 -s -- `
Run another  basic monitor in shared more  (or any other NF really)
`./go.sh -l 0 -- -r 1 -s -- `
*Then Ctrl-C the first NF (the one not running in shared mode)
The second NF should now be reallocated to the core that the first NF was running on. 

| This PR includes         |          |
| ------------------------ | -------- |
| Resolves issues          |  
| Breaking API changes     |
| Internal API changes     |
| Usability improvements   |  
| Bug fixes                |
| New functionality        | :+1: 
| New NF/onvm_mgr args     | 
| Changes to starting NFs  |  
| Dependency updates       | 
| Web stats updates        | 

## Merging notes:
 - Dependencies: None  

**TODO before merging :**
 - [x] PR is ready for review

## Test Plan:
Figure out if this can break anything.

### Review checklist:  

- Sanity checks, assigned to @dennisafa  @rskennedy 
    - Run make in `/onvm` and `/examples` directories
    - Test the new functionality

- Code style, assigned to @dennisafa  @rskennedy 
    - Run linter
    - Check everything style related

- Code design, assigned to @dennisafa  @rskennedy 
    - Check for memory leaks
    - Check that code is reusable
    - Code is clean, functions are concise
    - Verify that edge cases properly handled

- Performance, assigned to @dennisafa  @rskennedy 
    - Run Speed Tester NF, report performance.
    - Run pktgen, report performance (not really required)

- Documentation, assigned to @dennisafa  @rskennedy 
    - Check if the new changes are well documented, in both code and READMEs

### CODE REVIEW

**Comment by onvm** [comment]
> In response to PR creation

## CI Message

Your results will arrive shortly

**Comment by onvm** [comment]
> In response to PR creation

## CI Message

Run successful see results:
[Results from nimbnode30]
Median TX pps for Speed Tester: 34988371

## Linter Failed

examples/aes_decrypt/aes.c:523:  If an else has a brace on one side, it should have it on both  [readability/braces] [5]
examples/aes_decrypt/aes.c:1024:  If an else has a brace on one side, it should have it on both  [readability/braces] [5]
examples/aes_decrypt/aes.c:1028:  If an else has a brace on one side, it should have it on both  [readability/braces] [5]
examples/aes_decrypt/aes.c:1081:  If an else has a brace on one side, it should have it on both  [readability/braces] [5]
examples/aes_decrypt/aes.c:1087:  If an else has a brace on one side, it should have it on both  [readability/braces] [5]
Total errors found: 5
examples/aes_encrypt/aes.c:523:  If an else has a brace on one side, it should have it on both  [readability/braces] [5]
examples/aes_encrypt/aes.c:1024:  If an else has a brace on one side, it should have it on both  [readability/braces] [5]
examples/aes_encrypt/aes.c:1028:  If an else has a brace on one side, it should have it on both  [readability/braces] [5]
examples/aes_encrypt/aes.c:1081:  If an else has a brace on one side, it should have it on both  [readability/braces] [5]
examples/aes_encrypt/aes.c:1087:  If an else has a brace on one side, it should have it on both  [readability/braces] [5]
Total errors found: 5
examples/arp_response/arp_response.c:144:  Missing space before ( in switch(  [whitespace/parens] [5]
Total errors found: 1
examples/bridge/bridge.c:135:  Missing space before ( in if(  [whitespace/parens] [5]
examples/bridge/bridge.c:138:  If an else has a brace on one side, it should have it on both  [readability/braces] [5]
examples/bridge/bridge.c:156:  If an else has a brace on one side, it should have it on both  [readability/braces] [5]
Total errors found: 3
examples/flow_table/flow_table.c:103:  Missing space before ( in if(  [whitespace/parens] [5]
examples/flow_table/flow_table.c:198:  Missing space before ( in if(  [whitespace/parens] [5]
examples/flow_table/flow_table.c:216:  Missing space before ( in if(  [whitespace/parens] [5]
examples/flow_table/flow_table.c:224:  Missing space before ( in if(  [whitespace/parens] [5]
examples/flow_table/flow_table.c:231:  If an else has a brace on one side, it should have it on both  [readability/braces] [5]
examples/flow_table/flow_table.c:238:  If an else has a brace on one side, it should have it on both  [readability/braces] [5]
Total errors found: 6
examples/flow_table/msgbuf.c:71:  Missing space before ( in if(  [whitespace/parens] [5]
examples/flow_table/msgbuf.c:73:  Missing space before ( in if(  [whitespace/parens] [5]
examples/flow_table/msgbuf.c:83:  Missing space before ( in while(  [whitespace/parens] [5]
examples/flow_table/msgbuf.c:86:  Missing space before ( in if(  [whitespace/parens] [5]
examples/flow_table/msgbuf.c:89:  Missing space before ( in if(  [whitespace/parens] [5]
examples/flow_table/msgbuf.c:106:  Missing space before ( in if(  [whitespace/parens] [5]
examples/flow_table/msgbuf.c:108:  Missing space before ( in if(  [whitespace/parens] [5]
examples/flow_table/msgbuf.c:116:  Missing space before ( in while(  [whitespace/parens] [5]
examples/flow_table/msgbuf.c:119:  Missing space before ( in if(  [whitespace/parens] [5]
examples/flow_table/msgbuf.c:125:  Missing space before ( in if(  [whitespace/parens] [5]
examples/flow_table/msgbuf.c:138:  Extra space before last semicolon. If this should be an empty statement, use {} instead.  [whitespace/semicolon] [5]
examples/flow_table/msgbuf.c:140:  Missing space before ( in if(  [whitespace/parens] [5]
examples/flow_table/msgbuf.c:149:  Missing space before ( in if(  [whitespace/parens] [5]
examples/flow_table/msgbuf.c:157:  Missing space before ( in if(  [whitespace/parens] [5]
examples/flow_table/msgbuf.c:159:  Missing space before ( in if(  [whitespace/parens] [5]
examples/flow_table/msgbuf.c:162:  Missing space before ( in if(  [whitespace/parens] [5]
examples/flow_table/msgbuf.c:169:  Missing space before ( in while(  [whitespace/parens] [5]
Total errors found: 17
examples/flow_table/sdn.c:99:  Extra space before last semicolon. If this should be an empty statement, use {} instead.  [whitespace/semicolon] [5]
examples/flow_table/sdn.c:125:  Missing space before ( in if(  [whitespace/parens] [5]
examples/flow_table/sdn.c:127:  Missing space before ( in if(  [whitespace/parens] [5]
examples/flow_table/sdn.c:147:  If an else has a brace on one side, it should have it on both  [readability/braces] [5]
examples/flow_table/sdn.c:147:  Missing space before {  [whitespace/braces] [5]
examples/flow_table/sdn.c:153:  Missing space before ( in while(  [whitespace/parens] [5]
examples/flow_table/sdn.c:160:  Missing space before {  [whitespace/braces] [5]
examples/flow_table/sdn.c:235:  If an else has a brace on one side, it should have it on both  [readability/braces] [5]
examples/flow_table/sdn.c:239:  If an else has a brace on one side, it should have it on both  [readability/braces] [5]
examples/flow_table/sdn.c:279:  Missing space before ( in if(  [whitespace/parens] [5]
examples/flow_table/sdn.c:418:  Missing space before ( in if(  [whitespace/parens] [5]
examples/flow_table/sdn.c:423:  Missing space before ( in if(  [whitespace/parens] [5]
examples/flow_table/sdn.c:493:  If an else has a brace on one side, it should have it on both  [readability/braces] [5]
examples/flow_table/sdn.c:501:  If an else has a brace on one side, it should have it on both  [readability/braces] [5]
examples/flow_table/sdn.c:538:  Missing space before ( in if(  [whitespace/parens] [5]
examples/flow_table/sdn.c:542:  Missing space before ( in if(  [whitespace/parens] [5]
examples/flow_table/sdn.c:545:  Missing space before ( in if(  [whitespace/parens] [5]
examples/flow_table/sdn.c:547:  Missing space before ( in if(  [whitespace/parens] [5]
examples/flow_table/sdn.c:555:  Missing space before ( in if(  [whitespace/parens] [5]
examples/flow_table/sdn.c:569:  Missing space before ( in while(  [whitespace/parens] [5]
Total errors found: 20
examples/flow_table/setupconn.c:82:  Missing space before ( in if(  [whitespace/parens] [5]
examples/flow_table/setupconn.c:89:  Missing space before ( in if(  [whitespace/parens] [5]
examples/flow_table/setupconn.c:94:  Missing space before ( in if(  [whitespace/parens] [5]
examples/flow_table/setupconn.c:102:  Missing space before ( in if(  [whitespace/parens] [5]
examples/flow_table/setupconn.c:110:  Missing space before {  [whitespace/braces] [5]
examples/flow_table/setupconn.c:112:  Missing space before {  [whitespace/braces] [5]
examples/flow_table/setupconn.c:122:  Missing space before ( in if(  [whitespace/parens] [5]
examples/flow_table/setupconn.c:140:  Missing space before ( in if(  [whitespace/parens] [5]
examples/flow_table/setupconn.c:149:  Missing space before ( in if(  [whitespace/parens] [5]
examples/flow_table/setupconn.c:156:  Missing space before ( in if(  [whitespace/parens] [5]
Total errors found: 10
examples/flow_tracker/flow_tracker.c:108:  Missing space before ( in switch(  [whitespace/parens] [5]
examples/flow_tracker/flow_tracker.c:210:  If an else has a brace on one side, it should have it on both  [readability/braces] [5]
Total errors found: 2
examples/scaling_example/scaling.c:312:  Missing space before ( in while(  [whitespace/parens] [5]
Total errors found: 1
onvm/onvm_mgr/main.c:98:  Mismatching spaces inside () in while  [whitespace/parens] [5]
Total errors found: 1
onvm/onvm_mgr/onvm_args.c:166:  Missing space before ( in if(  [whitespace/parens] [5]
onvm/onvm_mgr/onvm_args.c:166:  Missing space before {  [whitespace/braces] [5]
onvm/onvm_mgr/onvm_args.c:172:  Missing space before ( in if(  [whitespace/parens] [5]
onvm/onvm_mgr/onvm_args.c:172:  Missing space before {  [whitespace/braces] [5]
onvm/onvm_mgr/onvm_args.c:328:  Missing space before ( in if(  [whitespace/parens] [5]
onvm/onvm_mgr/onvm_args.c:352:  Missing space before {  [whitespace/braces] [5]
onvm/onvm_mgr/onvm_args.c:357:  Missing space before ( in if(  [whitespace/parens] [5]
Total errors found: 7
onvm/onvm_mgr/onvm_init.c:222:  Never use sprintf. Use snprintf instead.  [runtime/printf] [5]
Total errors found: 1
onvm/onvm_mgr/onvm_nf.c:192:  Missing space before ( in if(  [whitespace/parens] [5]
onvm/onvm_mgr/onvm_nf.c:266:  Missing space before ( in if(  [whitespace/parens] [5]
Total errors found: 2
onvm/onvm_mgr/onvm_stats.c:186:  Missing space before ( in if(  [whitespace/parens] [5]
onvm/onvm_mgr/onvm_stats.c:186:  Missing space before {  [whitespace/braces] [5]
Total errors found: 2
onvm/onvm_nflib/onvm_config_common.c:52:  Missing space before {  [whitespace/braces] [5]
onvm/onvm_nflib/onvm_config_common.c:112:  Missing space before ( in for(  [whitespace/parens] [5]
onvm/onvm_nflib/onvm_config_common.c:112:  Missing space before {  [whitespace/braces] [5]
onvm/onvm_nflib/onvm_config_common.c:417:  Never use sprintf. Use snprintf instead.  [runtime/printf] [5]
onvm/onvm_nflib/onvm_config_common.c:433:  Never use sprintf. Use snprintf instead.  [runtime/printf] [5]
onvm/onvm_nflib/onvm_config_common.c:487:  Never use sprintf. Use snprintf instead.  [runtime/printf] [5]
Total errors found: 6
onvm/onvm_nflib/onvm_flow_dir.c:64:  Missing space before ( in if(  [whitespace/parens] [5]
onvm/onvm_nflib/onvm_flow_dir.c:95:  Missing space before {  [whitespace/braces] [5]
onvm/onvm_nflib/onvm_flow_dir.c:103:  Missing space before {  [whitespace/braces] [5]
onvm/onvm_nflib/onvm_flow_dir.c:111:  Missing space before {  [whitespace/braces] [5]
onvm/onvm_nflib/onvm_flow_dir.c:128:  Missing space before {  [whitespace/braces] [5]
onvm/onvm_nflib/onvm_flow_dir.c:143:  Missing space before {  [whitespace/braces] [5]
onvm/onvm_nflib/onvm_flow_dir.c:151:  Missing space before {  [whitespace/braces] [5]
onvm/onvm_nflib/onvm_flow_dir.c:159:  Missing space before {  [whitespace/braces] [5]
onvm/onvm_nflib/onvm_flow_dir.c:176:  Missing space before {  [whitespace/braces] [5]
Total errors found: 9
onvm/onvm_nflib/onvm_nflib.c:341:  Missing space before ( in if(  [whitespace/parens] [5]
onvm/onvm_nflib/onvm_nflib.c:344:  Missing space before ( in if(  [whitespace/parens] [5]
onvm/onvm_nflib/onvm_nflib.c:347:  Missing space before ( in if(  [whitespace/parens] [5]
onvm/onvm_nflib/onvm_nflib.c:351:  Missing space before ( in if(  [whitespace/parens] [5]
onvm/onvm_nflib/onvm_nflib.c:354:  Missing space before ( in if(  [whitespace/parens] [5]
onvm/onvm_nflib/onvm_nflib.c:357:  Missing space before ( in if(  [whitespace/parens] [5]
onvm/onvm_nflib/onvm_nflib.c:584:  Missing space before ( in switch(  [whitespace/parens] [5]
onvm/onvm_nflib/onvm_nflib.c:750:  Missing space before ( in if(  [whitespace/parens] [5]
onvm/onvm_nflib/onvm_nflib.c:761:  Missing space before ( in if(  [whitespace/parens] [5]
onvm/onvm_nflib/onvm_nflib.c:828:  Missing space before {  [whitespace/braces] [5]
Total errors found: 10
onvm/onvm_nflib/onvm_pkt_helper.c:313:  Missing space before ( in switch(  [whitespace/parens] [5]
onvm/onvm_nflib/onvm_pkt_helper.c:376:  Missing space before {  [whitespace/braces] [5]
Total errors found: 2

**Comment by koolzz** [comment]
PR review assigned to  @dennisafa  @rskennedy

**Comment by koolzz** [comment]
@onvm hello my old friend

**Comment by onvm** [comment]
> @onvm hello my old friend

## CI Message

Your results will arrive shortly

**Comment by onvm** [comment]
> @onvm hello my old friend

## CI Message

Run successful see results:
[Results from nimbnode30]
Median TX pps for Speed Tester: 34822438

## Linter Passed

File: onvm/onvm_nflib/onvm_nflib.c [hunk 1]
```diff
@@ -585,7 +585,14 @@ onvm_nflib_handle_msg(struct onvm_nf_msg *msg, __attribute__((unused)) struct on
                         RTE_LOG(INFO, APP, "Received scale message...\n");
                         onvm_nflib_scale((struct onvm_nf_scale_info *)msg->msg_data);
                         break;
-                case MSG_NOOP:
+                case MSG_CHANGE_CORE:
+                        RTE_LOG(INFO, APP, "Recieved relocation message...\n");
```
Comment (dennisafa) [inline onvm/onvm_nflib/onvm_nflib.c]: typo: Received

File: onvm/onvm_nflib/onvm_threading.c [hunk 2]
```diff
@@ -137,3 +137,41 @@ onvm_threading_core_affinitize(int cpu) {
 
         return rte_thread_set_affinity(&cpus);
 }
+
+int
+onvm_threading_find_nf_to_reassign_core(uint16_t candidate_core, struct core_status *cores) {
+        int i;
+        int candidate_nf_id, most_used_core;
+        int max_nfs_per_core;
```
Comment (dennisafa) [inline onvm/onvm_nflib/onvm_threading.c]: maybe combine all these into one?
int i, candidate_nf_id, most_used_core, max_nfs_per_core;

File: onvm/onvm_nflib/onvm_threading.h [hunk 3]
```diff
@@ -84,4 +90,19 @@ onvm_threading_get_core(uint16_t *core_value, uint8_t flags, struct core_status
...
+ * @param cores
+ *    A pointer to the core_status map containing core information
+ *
+ * @return
+ *    An instance ID of a candidate NF that should be moved to candidate_core
+ */
+int
+onvm_threading_find_nf_to_reassign_core(uint16_t candidate_core, struct core_status *cores);
+
+
```
Comment (dennisafa) [inline onvm/onvm_nflib/onvm_threading.h]: style nit: remove line

File: onvm/onvm_mgr/onvm_nf.c [hunk 4]
```diff
@@ -84,6 +84,17 @@ onvm_nf_ready(struct onvm_nf_info *nf_info);
...
 
+/*
+ * Function to move a NF to another core.
+ *
+ * Input  : instance id of the NF that needs to be moved
+ *          new_core value of where the NF should be moved
+ * Output : an error code
+ *
+ */
+inline int
```
Comment (dennisafa) [inline onvm/onvm_mgr/onvm_nf.c]: 95:  Line ends in whitespace.

File: onvm/onvm_mgr/onvm_nf.c [hunk 5]
```diff
@@ -285,5 +300,30 @@ onvm_nf_stop(struct onvm_nf_info *nf_info) {
...
+        if (candidate_nf_id > 0) {
+                onvm_nf_relocate_nf(candidate_nf_id, candidate_core);
+        }
+
+        return 0;
+}
+
+inline int onvm_nf_relocate_nf(uint16_t dest, uint16_t new_core) {
+        uint16_t *msg_data;
+
```
Comment (dennisafa) [inline onvm/onvm_mgr/onvm_nf.c]: 314:  Line ends in whitespace.

File: onvm/onvm_nflib/onvm_nflib.c [hunk 6]
```diff
@@ -585,7 +585,14 @@ onvm_nflib_handle_msg(struct onvm_nf_msg *msg, __attribute__((unused)) struct on
                         RTE_LOG(INFO, APP, "Received scale message...\n");
                         onvm_nflib_scale((struct onvm_nf_scale_info *)msg->msg_data);
                         break;
-                case MSG_NOOP:
+                case MSG_CHANGE_CORE:
+                        RTE_LOG(INFO, APP, "Received relocation message...\n");
```
Comment (dennisafa) [inline onvm/onvm_nflib/onvm_nflib.c]: 589:  Line ends in whitespace.

File: onvm/onvm_nflib/onvm_nflib.c [hunk 7]
```diff
@@ -585,7 +585,14 @@ onvm_nflib_handle_msg(struct onvm_nf_msg *msg, __attribute__((unused)) struct on
                         RTE_LOG(INFO, APP, "Received scale message...\n");
                         onvm_nflib_scale((struct onvm_nf_scale_info *)msg->msg_data);
                         break;
-                case MSG_NOOP:
+                case MSG_CHANGE_CORE:
+                        RTE_LOG(INFO, APP, "Received relocation message...\n"); 
+                        RTE_LOG(INFO, APP, "Moving NF to core %d\n", *(uint16_t *)msg->msg_data);
```
Comment (dennisafa) [inline onvm/onvm_nflib/onvm_nflib.c]: 590:  Line ends in whitespace.

**Comment by koolzz** [comment]
@dennisafa implemented requested changes, approve if everything else is good.

**Comment by koolzz** [comment]
@rskennedy give this a review when you can

**Comment by koolzz** [comment]
The more I though about this we should create a macro for it. This is optimization that might not always be wanted.

**Comment by koolzz** [comment]
@onvm updated.

**Comment by onvm** [comment]
> @onvm updated.

## CI Message

Your results will arrive shortly

**Comment by onvm** [comment]
> @onvm updated.

## CI Message

Run successful see results:
[Results from nimbnode30]
Median TX pps for Speed Tester: 35204966

## Linter Passed

**Comment by koolzz** [comment]
@dennisafa does introducing the macro make sense?

**Comment by dennisafa** [comment]
> @dennisafa does introducing the macro make sense?

Adding a macro makes sense. 

> @dennisafa does introducing the macro make sense?

Yea, I'm a fan of the macros. It makes sense to be able to enable or disable them (like ONVM_NF_SHUTDOWN_CORE_REASSIGNMENT)  based on what your  purposes may be.

**Comment by koolzz** [comment]
Will update with the new config approach when shared cpu changes are merged

**Comment by koolzz** [comment]
Don't merge still need changes from the #117

**Comment by kevindweb** [comment]
What needs to be done to merge this? #117 is merged. Do you want me to pull develop and fix the changes, and submit a pr to you?

**Comment by koolzz** [comment]
Nah I'll do it don't worry, it's on my list. I was planning to put the macro into the onvm config struct

**Comment by koolzz** [comment]
This is a pretty small core managing feature that isn't that useful on its own especially as its disabled by default. I feel like we can merge this later there is no rush to squeeze it into 19.05

**Comment by koolzz** [comment]
wait why did I close this whoops

**Comment by kevindweb** [comment]
@onvm should be put in queue

**Comment by onvm** [comment]
> @onvm should be put in queue

## CI Message

Another CI run in progress, adding request to the end of the list

**Comment by kevindweb** [comment]
@onvm hopefully this doesn't run

**Comment by onvm** [comment]
> @onvm hopefully this doesn't run

## CI Message

Duplicate request already waiting, ignoring message

**Comment by onvm** [comment]
> @onvm should be put in queue

## CI Message

Your results will arrive shortly

**Comment by kevindweb** [comment]
@onvm not sure why nfd isn't working

**Comment by onvm** [comment]
> @onvm not sure why nfd isn't working

## CI Message

Your results will arrive shortly

**Comment by onvm** [comment]
> @onvm not sure why nfd isn't working

## CI Message

Error: ERROR: Failed to post results to GitHub

**Comment by kevindweb** [comment]
@onvm this should be better now

**Comment by onvm** [comment]
> @onvm this should be better now

## CI Message

Your results will arrive shortly

**Comment by onvm** [comment]
> @onvm this should be better now

## CI Message

Error: ERROR: Failed to post results to GitHub

**Comment by kevindweb** [comment]
@onvm test

**Comment by kevindweb** [comment]
@onvm now really test

**Comment by onvm** [comment]
> @onvm now really test

## CI Message

Your results will arrive shortly

**Comment by onvm** [comment]
> @onvm now really test

## CI Message

Error: ERROR: Failed to fetch results from nimbnode17

**Comment by kevindweb** [comment]
@onvm please send 0

**Comment by onvm** [comment]
> @onvm please send 0 

## CI Message

Your results will arrive shortly

**Comment by koolzz** [comment]
@onvm 
Updated with the latest version

**Comment by onvm** [comment]
> Testing

## CI Message

Your results will arrive shortly

**Comment by kevindweb** [comment]
I'll fix that mTCP thing (not a bug apache benchmark just went faster), I'm just trying to make sure pktgen is working

**Comment by onvm** [comment]
> Testing

## CI Message

Error: ERROR: Failed to copy ONVM files to nimbnode17

**Comment by onvm** [comment]
> Testing

## CI Message

Error: ERROR: Script failed on nimbnode17

**Pull request merged**